File: SO/04-14/ProducerConsumer.java
// ProducerConsumer.java
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

class Producer implements Runnable {
    private BlockingQueue<Integer> queue;

    public Producer(BlockingQueue<Integer> queue) {
        this.queue = queue;
    }

    public void run() {
        try {
            while(true) {
                int item = (int)(Math.random() * 100);
                queue.put(item);
                System.out.println("Producer produced: " + item);
                Thread.sleep(1000);
            }
        } catch(InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

class Consumer implements Runnable {
    private BlockingQueue<Integer> queue;

    public Consumer(BlockingQueue<Integer> queue) {
        this.queue = queue;
    }

    public void run() {
        try {
            while(true) {
                int item = queue.take();
                System.out.println("Consumer consumed: " + item);
                Thread.sleep(2000);
            }
        } catch(InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

public class ProducerConsumer {
    public static void main(String[] args) {
        BlockingQueue<Integer> queue = new ArrayBlockingQueue<>(5);
        Thread producer = new Thread(new Producer(queue));
        Thread consumer = new Thread(new Consumer(queue));
        producer.start();
        consumer.start();
    }
}

File: SO/04-14/ProducerConsumer.kt
// File: ProducerConsumer.kt
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel

fun main() = runBlocking {
    val channel = Channel<Int>(5)

    launch {
        for (i in 0 until 10) {
            channel.send(i)
            println("Producer produced: $i")
            delay(1000L)
        }
        channel.close()
    }

    launch {
        for (item in channel) {
            println("Consumer consumed: $item")
            delay(2000L)
        }
    }
}
